cpp/src/arrow/c/bridge.cc

```diff
@@ -586,8 +586,12 @@ struct ArrayExporter {
       ++buffers_begin;
     }
 
+    Type::type storage_id =
+        data->type->id() == Type::EXTENSION
+            ? (checked_cast<const ExtensionType&>(*data->type).storage_type()->id())
+            : data->type->id();
     bool need_variadic_buffer_sizes =
-        data->type->id() == Type::BINARY_VIEW || data->type->id() == Type::STRING_VIEW;
+        storage_id == Type::BINARY_VIEW || storage_id == Type::STRING_VIEW;
```

Comment:
There's a special accessor for this https://github.com/apache/arrow/blob/219d207f97152a5017cc61c0f925dc4387adc6f7/cpp/src/arrow/type.h#L184

```suggestion
        data->type->storage_id() == Type::BINARY_VIEW || storage_id == Type::STRING_VIEW;
```